gtsam/linear/LinearSolver.cpp

```diff
@@ -0,0 +1,45 @@
+/* ----------------------------------------------------------------------------
+
+ * GTSAM Copyright 2010, Georgia Tech Research Corporation,
+ * Atlanta, Georgia 30332-0415
+ * All Rights Reserved
+ * Authors: Frank Dellaert, et al. (see THANKS for the full author list)
+
+ * See LICENSE for the license information
+
+ * -------------------------------------------------------------------------- */
+
+/**
+ * @file    LinearSolver.cpp
+ * @brief   Common Interface for Linear Solvers
+ * @author  Fan Jiang
+ */
+
+#include <gtsam/linear/LinearSolver.h>
+#include <gtsam/linear/SparseEigenSolver.h>
+#include <gtsam/linear/SuiteSparseSolver.h>
+#include <gtsam/linear/CuSparseSolver.h>
+
+namespace gtsam {
+
+LinearSolver::LinearSolver() = default;
+
+boost::shared_ptr<LinearSolver> LinearSolver::fromLinearSolverParams(
```

Comment:
I'm a little bit confused why the elimination methods aren't also considered LinearSolver's... maybe that doesn't make sense but would you mind clarifying how elimination methods and linear solvers are implemented with different APIs?

I guess the elimination methods kind of get special treatment by being implemented in the GaussianFactorGraph and related templated BayesNet, BayesTree, etc classes, whereas the other linear solving methods do not get this special treatment and are instead fully self-contained in their own classes.

As a follow-up, is it possible to add some solvers to Marginals?  For example, Ceres can generate covariances - could those be cheaply converted to some sort of bayes tree form?  Or in sparse QR convert the sparse R matrix to Bayes tree form?

Reply:
They are but problem is, they are deeply entangled with the GaussianFactorGraph etc etc that it's very hard to refactor them "out of" GFG. I tried a year ago but without much success.